test: Pass ArgsManager into getarg_tests

glowang opened the pull request:
Replaced the global argsManager gArgs with a locally defined one in getarg_tests. This is to avoid confusion in arg settings between the test's ArgsManager and the   #18804

glowang committed:
Create a local class inherited from BasicTestingSetup with a localized args manager
and put it into the getarg_tests namespace

ryanofsky commented on src/test/getarg_tests.cpp:
```diff
@@ -30,7 +30,7 @@ static void ResetArgs(const std::string& strArg)
 
     std::string error;
-    BOOST_CHECK(gArgs.ParseParameters(vecChar.size(), vecChar.data(), error));
+    BOOST_CHECK(argsman.ParseParameters(vecChar.size(), vecChar.data(), error));
```
In commit "scripted-diff: replace gArgs with a local args manager" (af069fcdd9b5e2c983ee32f0292c8e1f167359ec)

In the commit message for https://github.com/bitcoin/bitcoin/commit/af069fcdd9b5e2c983ee32f0292c8e1f167359ec.patch you need an extra newline between the subject and body of the commit for the scripted diff to work like:

```
scripted-diff: replace gArgs with a local args manager

-BEGIN VERIFY SCRIPT-
sed -i -e '205!s/gArgs/argsman/g' src/test/getarg_tests.cpp
-END VERIFY SCRIPT-
```

glowang replied:
Thank you! I will update the developer README too then.

ryanofsky commented on src/test/getarg_tests.cpp:
```diff
@@ -33,6 +33,8 @@ static void ResetArgs(const std::string& strArg)
     BOOST_CHECK(gArgs.ParseParameters(vecChar.size(), vecChar.data(), error));
 }
 
+ArgsManager argsman;
```
In commit "Include an unused ArgsManager" (2e0c74b1ffe9bd1867700fd4a58458160324e23c)

This variable is still shared between test instances (and I think also accesible outside the test file using `extern`). I think it would be better to add a member variable to the boost test setup so the tests are isolated. This would look something like 

https://github.com/bitcoin/bitcoin/blob/db762529d6bffcc6ab0c1c64b9378925f2d75fa1/src/test/getarg_tests.cpp#L16-L23

https://github.com/ryanofsky/bitcoin/commit/db762529d6bffcc6ab0c1c64b9378925f2d75fa1

glowang replied:
Thank you, will fix!

ryanofsky reviewed (approved):
Code review ACK af069fcdd9b5e2c983ee32f0292c8e1f167359ec. This change seems good and is an improvement, but I left a suggestion to use a unique ArgsManager for each test case instead of a shared one.

MarcoFalke commented on doc/developer-notes.md:
```diff
@@ -960,6 +960,7 @@ To create a scripted-diff:
 - in the commit message include the bash script between lines containing just the following text:
     - `-BEGIN VERIFY SCRIPT-`
     - `-END VERIFY SCRIPT-`
+- note: you need an extra newline between the above two steps for the scripted diff to work properly
```
The newline is generally needed for git to work properly. This is already explained in CONTRIBUTING, see this section:

```
Commit messages should be verbose by default consisting of a short subject line
(50 chars max), a blank line and detailed explanatory text as separate
paragraph(s), unless the title alone is self-explanatory (like "Corrected typo
in init.cpp") in which case a single title line is sufficient. Commit messages should be
helpful to people reading your code in the future, so explain the reasoning for
your decisions. Further explanation [here](https://chris.beams.io/posts/git-commit/).

maflcko reviewed (commented):
Concept ACK. Thanks for switching the tests away from using `gArgs` this could badly interfere with later tests and cause hard to debug run-time issues

glowang commented on src/test/getarg_tests.cpp:
```diff
@@ -13,10 +13,17 @@
 #include <boost/test/unit_test.hpp>
 
-BOOST_FIXTURE_TEST_SUITE(getarg_tests, BasicTestingSetup)
+class LocalTestingSetup : BasicTestingSetup {
```
I kept getting `cannot define or redeclare 'ResetArgs' here because namespace 'getarg_tests' does not enclose namespace
      'LocalTestingSetup'`

MarcoFalke replied:
>  I kept getting cannot define or redeclare 'ResetArgs' here because namespace 'getarg_tests' does not enclose namespace 'LocalTestingSetup'

You can put this into a namespace. Like in `validation_block_tests`

glowang commented on src/test/getarg_tests.cpp:
```diff
@@ -200,7 +200,7 @@ BOOST_AUTO_TEST_CASE(logargs)
     const auto okaylog_negbool = std::make_pair("-okaylog-negbool", ArgsManager::ALLOW_BOOL);
     const auto okaylog = std::make_pair("-okaylog", ArgsManager::ALLOW_ANY);
     const auto dontlog = std::make_pair("-dontlog", ArgsManager::ALLOW_ANY | ArgsManager::SENSITIVE);
```
> test/getarg_tests.cpp:218: error: in "getarg_tests/logargs": check str.find("Command-line arg: okaylog-bool=\"\"") != std::string::npos has failed
> test/getarg_tests.cpp:219: error: in "getarg_tests/logargs": check str.find("Command-line arg: okaylog-negbool=false") != std::string::npos has failed
> test/getarg_tests.cpp:220: error: in "getarg_tests/logargs": check str.find("Command-line arg: okaylog=\"public\"") != std::string::npos has failed
> test/getarg_tests.cpp:221: error: in "getarg_tests/logargs": check str.find("dontlog=****") != std::string::npos has failed

My tests passed in the first 2 commits, but this commit gives me these errors locally. It looks like the bug comes from my `debug.log` having something that it should not. My struggle is I can't really find my debug log...I checked out some resources (Andrew Chow's/bitcoin wiki), but none seems to mention where to locate this file. 

https://gist.github.com/fjahr/2cd23ad743a2ddfd4eed957274beca0f does include an example command `$ cat ~/Library/Application\ Support/Bitcoin/regtest/debug.log` but this doesn't work for me locally. Any tips? @amitiuttarwar   @fjahr  🙏

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/18926#discussion_r427038185

> My tests passed in the first 2 commits, but this commit gives me these errors locally. It looks like the bug comes from my `debug.log` having something that it should not.

That's right, the problem is happening because the arguments are being set in the `argman` instance not `gArgs` but line 214 still says `gArgs.LogArgs()` not `argsman.LogArgs()`.

> My struggle is I can't really find my debug log...I checked out some resources (Andrew Chow's/bitcoin wiki), but none seems to mention where to locate this file.

I believe it's written to a temporary directory so you could check TMPDIR, but most convenient way to see logs from tests is:

```
src/test/test_bitcoin -l test_suite -t getarg_tests -- DEBUG_LOG_OUT
```

Also, in my previous comment https://github.com/bitcoin/bitcoin/pull/18926#discussion_r423913184, I did link to a working updated copy of the test. If you want to apply and test those changes locally you can do that with:

```c++
git fetch https://github.com/ryanofsky/bitcoin review.18926.1-edit
git checkout FETCH_HEAD src/test/getarg_tests.cpp
make -C src test/test_bitcoin
src/test/test_bitcoin -l test_suite -t getarg_tests -- DEBUG_LOG_OUT
```

glowang replied:
ahhhh thank you! I avoided line 214 in my sed command, because it would modify `LogArgs()` to `Loargsman()`, but I overlooked the use of the global manager in there. I will look what sed command I can use to avoid changing only a specific location of line.

I feel like if I commited your solution, my scripted-diff commit on its own is still failing the test...this makes my commits depend on each other, instead of working modularly.

MarcoFalke replied:
You can use 

```
$ echo 'gArgs.LogArgs()' | sed -e 's/\<gArgs\>/____/g'
____.LogArgs()

ryanofsky replied:
> I feel like if I commited your solution

To be sure, suggestion was just to compare against for debugging, not to commit

glowang replied:
> You can use
> 
> ```
> $ echo 'gArgs.LogArgs()' | sed -e 's/\<gArgs\>/____/g'
> ____.LogArgs()
> ```

Thanks Marco, I tried running this but it is not changing anything from my end. Is there any setup I needed to do? Are you using gnu-sed? This is what I tried.
`sed -i -e 's/\<gArgs\>/argsman/g' src/test/getarg_tests.cpp` 

I did receive a similar suggestion from my [post](https://stackoverflow.com/questions/61957896/sed-global-substitution-how-to-ignore-a-particular-range-of-positions-in-one?stw=2) on S/O, but somehow mine is not working 😂

glowang replied:
> > I feel like if I commited your solution
> 
> To be sure, suggestion was just to compare against for debugging, not to commit

Totally. Thanks again for your suggestion :)

glowang replied:
> > You can use
> > ```
> > $ echo 'gArgs.LogArgs()' | sed -e 's/\<gArgs\>/____/g'
> > ____.LogArgs()
> > ```
> 
> Thanks Marco, I tried running this but it is not changing anything from my end. Is there any setup I needed to do? Are you using gnu-sed? This is what I tried.
> `sed -i -e 's/\<gArgs\>/argsman/g' src/test/getarg_tests.cpp`
> 
> I did receive a similar suggestion from my [post](https://stackoverflow.com/questions/61957896/sed-global-substitution-how-to-ignore-a-particular-range-of-positions-in-one?stw=2) on S/O, but somehow mine is not working 😂

Maybe I should look more into the regex itself actually

MarcoFalke replied:
Yes, you will need to use gsed. Our scripted diffs don't support macos sed, only Ubuntu sed.

dongcarl commented on src/test/getarg_tests.cpp:
```diff
@@ -13,10 +13,19 @@
-BOOST_FIXTURE_TEST_SUITE(getarg_tests, BasicTestingSetup)
+namespace getarg_tests{
+    class LocalTestingSetup : BasicTestingSetup {
+        protected: 
+        void SetupArgs(ArgsManager& m_args, const std::vector<std::pair<std::string, unsigned int>>& args); 
+        void ResetArgs(const std::string& strArg); 
```
Might wanna remove these trailing whitespaces to appease Travis

glowang replied:
Thanks Carl 🙏

glowang commented:
This is ready for re-review.

MarcoFalke commented on src/test/getarg_tests.cpp:
```diff
@@ -13,9 +13,18 @@
+        protected:
+        void SetupArgs(ArgsManager& m_args, const std::vector<std::pair<std::string, unsigned int>>& args);
+        void ResetArgs(const std::string& strArg);
+        ArgsManager argsman;
```
```suggestion
        ArgsManager m_args;
```

according to the dev notes on member naming

MarcoFalke commented on src/test/getarg_tests.cpp:
```diff
@@ -13,9 +13,18 @@
+namespace getarg_tests{
+    class LocalTestingSetup : BasicTestingSetup {
+        protected:
+        void SetupArgs(ArgsManager& m_args, const std::vector<std::pair<std::string, unsigned int>>& args);
```
```suggestion
        void SetupArgs(const std::vector<std::pair<std::string, unsigned int>>& args);
```

This should be removed (along with the first commit), because in a member function you don't need to pass in member variables anymore

maflcko reviewed (commented):
Almost-ACK

glowang committed:
scripted-diff: replace gArgs with argsman

-BEGIN VERIFY SCRIPT-
sed -i 's/\<gArgs\>/m_args/g' src/test/getarg_tests.cpp
-END VERIFY SCRIPT-

MarcoFalke commented:
ACK f871f15c9d760b56a6a3b78b3941d3983d60810c

ryanofsky reviewed (approved):
Code review ACK f871f15c9d760b56a6a3b78b3941d3983d60810c. Changes look good and thanks for updating. In future would recommend using clang-format-diff and following [coding style](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c) notes, because it's atypical to indent namespace content, or indent protected keywords or put spaces around ::. Also it's fragile to define test setup class in a namespace, but test setup methods outside of the namespace and inside the test fixture instead. Would be simpler to just define the testing setup completely before using it without a namespace like: https://github.com/bitcoin/bitcoin/blob/8ad5f1c376fe99eeccac2696ac0e18e662323a4d/src/test/rpc_tests.cpp#L23 and it would have been a slightly smaller change too.

No need to change anything though, this is already better
